File: memories/2024/exercise/kotlin/day10/src/main/kotlin/delivery/Building4.kt
package delivery

/**
 * Building floor calculator for delivery instructions.
 */
object Building4 {
    private const val ELF_UP = 3
    private const val ELF_DOWN = -2
    private const val STANDARD_UP = 1
    private const val STANDARD_DOWN = -1
    private const val ELF_INDICATOR = "🧝"

    fun whichFloor(instructions: String): Int {
        val isElfMode = instructions.contains(ELF_INDICATOR)
        
        return instructions
            .filter { it == '(' || it == ')' }
            .sumOf { calculateFloorMovement(it, isElfMode) }
    }
    
    private fun calculateFloorMovement(direction: Char, isElfMode: Boolean): Int {
        return if (isElfMode) {
            if (direction == ')') ELF_UP else ELF_DOWN
        } else {
            if (direction == '(') STANDARD_UP else STANDARD_DOWN
        }
    }
}

File: memories/2024/exercise/ts/day10/src/delivery/building4.ts
/**
 * Building floor calculator for delivery instructions.
 */
export class Building4 {
    private static readonly ELF_UP = 3;
    private static readonly ELF_DOWN = -2;
    private static readonly STANDARD_UP = 1;
    private static readonly STANDARD_DOWN = -1;
    private static readonly ELF_INDICATOR = "🧝";

    static whichFloor(instructions: string): number {
        const isElfMode = instructions.includes(this.ELF_INDICATOR);
        
        return instructions
            .split('')
            .filter(ch => ch === '(' || ch === ')')
            .reduce((floor, ch) => floor + this.calculateFloorMovement(ch, isElfMode), 0);
    }
    
    private static calculateFloorMovement(direction: string, isElfMode: boolean): number {
        if (isElfMode) {
            return direction === ')' ? this.ELF_UP : this.ELF_DOWN;
        } else {
            return direction === '(' ? this.STANDARD_UP : this.STANDARD_DOWN;
        }
    }
}